File: backend/app/schemas/transfer.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict


class StegoImageSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    fragment_id: str
    carrier_id: str
    storage_provider: str
    storage_path: str


class TransferResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    original_filename: str
    fragment_count: int
    status: str
    processing_time_ms: float | None
    created_at: datetime


class FragmentSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    fragment_index: int
    size: int
    sha256: str


class StegoImageDetail(BaseModel):
    id: str
    fragment_id: str
    fragment_index: int
    carrier_id: str
    carrier_filename: str
    psnr_db: float | None
    ssim: float | None
    capacity_utilization: float | None


class TransferDetailResponse(TransferResponse):
    encrypted_file_id: str | None
    encryption_time_ms: float | None
    fragmentation_time_ms: float | None
    embedding_time_ms: float | None
    extraction_time_ms: float | None
    recovery_time_ms: float | None
    recovered: bool
    fragments: list[FragmentSummary]
    stego_images: list[StegoImageDetail]


class TransferHideResponse(BaseModel):
    transfer_id: str
    fragment_count: int
    selected_carrier_ids: list[str]
    stego_images: list[StegoImageSummary]
    processing_time_ms: float
    status: str


class TransferRecoverResponse(BaseModel):
    transfer_id: str
    status: str
    original_filename: str
    recovered_size: int
    integrity_verified: bool
    processing_time_ms: float

File: frontend/src/types/transfer.ts
export interface StegoImageSummary {
  id: string;
  fragment_id: string;
  carrier_id: string;
  storage_provider: string;
  storage_path: string;
}

export interface TransferHideResponse {
  transfer_id: string;
  fragment_count: number;
  selected_carrier_ids: string[];
  stego_images: StegoImageSummary[];
  processing_time_ms: number;
  status: string;
}

export interface TransferSummary {
  id: string;
  original_filename: string;
  fragment_count: number;
  status: string;
  processing_time_ms: number | null;
  created_at: string;
}

export interface FragmentSummary {
  id: string;
  fragment_index: number;
  size: number;
  sha256: string;
}

export interface StegoImageDetail {
  id: string;
  fragment_id: string;
  fragment_index: number;
  carrier_id: string;
  carrier_filename: string;
  psnr_db: number | null;
  ssim: number | null;
  capacity_utilization: number | null;
}

export interface TransferDetail extends TransferSummary {
  encrypted_file_id: string | null;
  encryption_time_ms: number | null;
  fragmentation_time_ms: number | null;
  embedding_time_ms: number | null;
  extraction_time_ms: number | null;
  recovery_time_ms: number | null;
  recovered: boolean;
  fragments: FragmentSummary[];
  stego_images: StegoImageDetail[];
}

export interface TransferRecoverResponse {
  transfer_id: string;
  status: string;
  original_filename: string;
  recovered_size: number;
  integrity_verified: boolean;
  processing_time_ms: number;
}
